src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileMetadataService.scala

```diff
@@ -62,38 +61,79 @@ class FileMetadataService(fileMetadataRepository: FileMetadataRepository,
                                       existingFileMetadataRows: Map[UUID, Seq[FilemetadataRow]]): Set[PropertyAction] = {
 
     val existingPropertiesForFile: Map[String, Seq[FilemetadataRow]] = existingFileMetadataRows.getOrElse(fileId, Seq()).groupBy(_.propertyname)
+    val existingPropertyNameAndItsValues: Map[String, Seq[String]] = existingPropertiesForFile.map {
+      case (propertyName, fileMetadataRow) => (propertyName, fileMetadataRow.map(_.value))
+    }
+
+    val metadataPropertiesGroupedByName: Set[((String, Boolean), Set[UpdateFileMetadataInput])] = metadataProperties.groupBy {
+      metadataProperty => (metadataProperty.filePropertyName, metadataProperty.filePropertyIsMultiValue)
+    }.toSet
+
+    metadataPropertiesGroupedByName.flatMap {
+      case ((metadataPropertyName, propertyIsMultiValue), propertyUpdates) =>
+        val fileMetadataRowsForProperty: Set[FilemetadataRow] = existingPropertiesForFile.getOrElse(metadataPropertyName, Set()).toSet
+        val existingValuesBelongingToProperty: Option[Seq[String]] = existingPropertyNameAndItsValues.get(metadataPropertyName)
+
+        generatePropertyActions(fileId, fileMetadataRowsForProperty, existingValuesBelongingToProperty, propertyUpdates, propertyIsMultiValue)
+    }
+  }
 
-    metadataProperties.map {
-      metadataProperty =>
-        if (!existingPropertiesForFile.contains(metadataProperty.filePropertyName)) {
-          PropertyAction("add", metadataProperty.filePropertyName, metadataProperty.value, fileId, uuidSource.uuid)
+  private def generatePropertyActions(fileId: UUID, fileMetadataRowsForProperty: Set[FilemetadataRow], existingValuesBelongingToProperty: Option[Seq[String]],
+                                      propertyUpdates: Set[UpdateFileMetadataInput], propertyIsMultiValue: Boolean): Set[PropertyAction] = {
+    existingValuesBelongingToProperty match {
+      case None => propertyUpdates.map(propertyUpdate => PropertyAction("add", propertyUpdate.filePropertyName, propertyUpdate.value, fileId, uuidSource.uuid))
+      case Some(valuesCurrentlyAssignedToProperty) =>
+        if (propertyIsMultiValue) {
+          val allNewValuesAreTheSame: Boolean = propertyUpdates.forall(propertyUpdate => valuesCurrentlyAssignedToProperty.contains(propertyUpdate.value))
+          val actionToPerform: String = if (allNewValuesAreTheSame) "noUpdate" else "delete"
+          val deleteOrNoUpdatePropertyActions: Set[PropertyAction] = fileMetadataRowsForProperty.map {
+            fileMetadataRow => PropertyAction(actionToPerform, fileMetadataRow.propertyname, fileMetadataRow.value, fileId, fileMetadataRow.metadataid)
+          }
+          val addPropertyActions: Set[_ <: PropertyAction] =
+            if (actionToPerform == "delete") { // delete the current fileMetadataRows and then add the new ones
+              generatePropertyActions(fileId, fileMetadataRowsForProperty, existingValuesBelongingToProperty = None, propertyUpdates, propertyIsMultiValue)
```

Comment:
Do you need the recursive call here. It's not tail recursion so the compiler isn't good at optimising it. It looks like the current call will hit the `case None` so could you replace the recursive call with `propertyUpdates.map(propertyUpdate => PropertyAction("add", propertyUpdate.filePropertyName, propertyUpdate.value, fileId, uuidSource.uuid))` ?
I'm not sure. I tried it and your tests still pass. What do you think?